### ISSUE TITLE
Typing issue with ValidationError.messages?

### ISSUE THREAD

**Issue by austinbutler** [marshmallow-code/marshmallow#1442]
Not sure if this is a documentation or typing issue, but `ValidationError.messages` is given as `Union[List[Any], Dict[Any, Any]]` while the docs say in a few places it's going to be a dictionary, at least when coming from `load()`.

For example, in the [validation](https://marshmallow.readthedocs.io/en/stable/quickstart.html#validation) documentation it says, "Schema.load() (and its JSON-decoding counterpart, Schema.loads()) raises a ValidationError error when invalid data are passed in. You can access the dictionary of validation errors from the ValidationError.messages attribute."

A slightly modified example from the documentation that shows it:

```python
from pprint import pprint

from marshmallow import Schema, fields, ValidationError


class BandMemberSchema(Schema):
    name = fields.String(required=True)
    email = fields.Email()


user_data = {"email": "mick@stones.com", "name": "Mick"}

try:
    BandMemberSchema().load(user_data)
    errors: dict = {}
except ValidationError as err:
    errors = err.messages
    pprint(err.messages)
```

Running it through MyPy:

```
> $ mypy m3example.py
m3example.py:17: error: Incompatible types in assignment (expression has type "Union[List[Any], Dict[Any, Any]]", variable has type "Dict[Any, Any]")
Found 1 error in 1 file (checked 1 source file)
```

- Python 3.7.4
- Marshmallow 3.2.1
- MyPy 0.740

Side note, the example in the docs gives MyPy errors as-is as well: `error: Argument 1 to "load" of "Schema" has incompatible type "List[Dict[str, str]]"; expected "Mapping[Any, Any]"`.

**Comment by deckar01** [comment]
The error messages value is guaranteed to be a dict for `Schema.load()`. The documentation is correct. The problem might be that mypy is keeping the type from `ValidationError`, which does allow lists, but only for field errors. All 3 of the places `Schema.load()` gets its `errors` value from are guaranteed to return a dict if the input is from a schema.

mypy should be able to tell that `ValidationError.normalized_messages()` always returns a dict, so maybe the problem is that `ErrorStore.errors` is getting its type from `ErrorStore.merge_errors()`, which also handles merging lists.

**Comment by deckar01** [comment]
I did some testing and mypy doesn't seem to care if the type of `errors` is dict. The signature of `ValidationError.messages` is the problem. Even if you wrap the value in dict constructor, it keeps the `Union` type when it stores the attribute.

This feely like a mypy bug. Here is a minimal repro (without marshmallow).

```py
from typing import Union, List, Dict

class Test:
    def __init__(self, data: Union[List, Dict]):
        self.data = data

data: dict = {}
test = Test(data)
data = test.data
# error: Incompatible types in assignment (expression has type "Union[List[Any], Dict[Any, Any]]", variable has type "Dict[Any, Any]")
```

Here is why I think this is a bug:

```py
from typing import Any

class Test:
    def __init__(self, data: Any):
        self.data = data

data: dict = {}
test = Test(data)
data = test.data
# Success: no issues found in 1 source file
```

This doesn't error even though we set the arg type to an even more generic type.

**Comment by deckar01** [comment]
While trying to trick mypy I found another bug:

```py
from typing import Union, List, Dict, Any

class Test:
    def __init__(self, data: Union[Dict, List]):
        if isinstance(data, List):
            self.data = data
        else:
            self.data = data

data: dict = {}
test = Test(data)
data = dict(test.data)
# error: Incompatible types in assignment (expression has type "Dict[Any, Any]", variable has type "List[Any]")
```

Regardless of how many branches there are, the type inference appears to always be the first branch regardless of the type of the value passed. Several other variations of the pattern confirmed this invalid behavior.

The only way I see to trick mypy into inferring the attribute is a dict, is to type check it as a dict first, before assignment. This would likely invalidate other type inference patterns, which are currently working correctly though.

**Comment by austinbutler** [comment]
Thanks, that at least clarifies the intent and confirms what I saw when I briefly peeked into the code.

For now I am working around it by casting `ValidationError.messages` to a dict in this case.

**Comment by sirosen** [comment]
I'm looking at old issues with the `typing` tag to see if we can clean anything up, especially as typing and mypy have matured a lot over the past couple of years.

The issue isn't really a `mypy` bug as far as I can tell. It's just the result of anything of the form:
```python
def f() -> list | dict:
    ...

x: dict = f()
```

`mypy` doesn't have a way to know that `err.messages` is a dict in this particular case.

I can think of two ways of "explaining the situation to type checkers".

It could tell if we added a new property like this to `ValidationError`:
```python
@property  # think of this as a "type asserting property"
def messages_dict(self) -> dict:
    if not isinstance(self.messages, dict):
        raise TypeError(
            f"cannot access 'messages_dict' when 'messages' is of type {type(self.messages)}"
        )
    return self.messages
```
or if we added a new subclass of `ValidationError` which guarantees that `messages` is a dict.

Do either of these appeal? I prefer the property if we're choosing one.
Otherwise, I would argue that this should be closed as a wontfix. `err.messages` is a union type and this is the behavior of unions.
